proof/refine/ARM/Finalise_R.thy

```diff
@@ -2329,6 +2329,118 @@ lemma replyUnlink_invs'[wp]:
   unfolding invs'_def valid_state'_def
   by wpsimp
 
+lemma setReply_list_refs_of_replies'[wp]:
+  "\<lbrace>\<lambda>s. P ((list_refs_of_replies' s)(p := list_refs_of_reply' r))\<rbrace>
+   setSchedContext p (scReply_update (\<lambda>_. replyPtrOpt) sc)
+   \<lbrace>\<lambda>rv s. P (list_refs_of_replies' s)\<rbrace>"
+  apply (wpsimp simp: setReply_def updateObject_default_def setObject_def split_def)
+  apply (erule arg_cong[where f=P, THEN iffD1, rotated])
+  apply (clarsimp simp: opt_map_def sym_refs_def fun_upd_def list_refs_of_reply'_def
+                        map_set_def projectKO_opt_reply)
+  apply (rule ext)
+  apply (clarsimp simp: projectKO_opt_reply list_refs_of_reply'_def)
+  oops
+
+crunches schedContextDonate
+  for replies_of'[wp]: "\<lambda>s. P (replies_of' s)"
+
+lemma isHeadSome:
+ "isHead (Some h) \<longleftrightarrow> (\<exists>p. h = Head p)"
+  by (clarsimp simp: isHead_def split: reply_next.split)
+
+lemma replyPrev_list_refs_of_replies':
+  "\<lbrakk>ko_at' reply replyPtr s; replyPrev reply = Some prevReply\<rbrakk>
+     \<Longrightarrow> (prevReply, ReplyPrev) \<in> list_refs_of_replies' s replyPtr"
+  by (clarsimp simp: obj_at'_def projectKOs list_refs_of_replies'_def opt_map_def
+                     list_refs_of_reply'_def
+              split: option.splits)
+
+lemma list_refs_of_replies'_eq:
+  "ko_at' reply replyPtr s \<Longrightarrow> list_refs_of_replies' s replyPtr = list_refs_of_reply' reply"
+  by (clarsimp simp: obj_at'_def projectKOs list_refs_of_replies'_def opt_map_def)
+
+lemma in_list_refs_of_replies'_E:
+  assumes in_list: "(p, tp) \<in> list_refs_of_replies' s replyPtr"
+  assumes nextR: "\<And>reply. tp = ReplyNext \<Longrightarrow> replyNext reply = Some (Next p)
+                            \<Longrightarrow> replies_of' s replyPtr = Some reply \<Longrightarrow> R"
+  assumes prevR: "\<And>reply. tp = ReplyPrev \<Longrightarrow> replyPrev reply = Some p
+                            \<Longrightarrow> replies_of' s replyPtr = Some reply \<Longrightarrow> R"
+  shows "R"
```

Comment:
This is a great lemma!
It's very nitpicky, but I would suggest reversing the order of some of the assumptions. It's equivalent, but where possible I like to introduce new variables before using them, to make it easier for humans to read.
```suggestion
lemma in_list_refs_of_replies'_E:
  assumes in_list: "(p, tp) \<in> list_refs_of_replies' s replyPtr"
  assumes nextR: "\<And>reply. tp = ReplyNext \<Longrightarrow> replies_of' s replyPtr = Some reply
                            \<Longrightarrow> replyNext reply = Some (Next p) \<Longrightarrow> R"
  assumes prevR: "\<And>reply. tp = ReplyPrev \<Longrightarrow> replies_of' s replyPtr = Some reply
                            \<Longrightarrow> replyPrev reply = Some p \<Longrightarrow> R"
  shows "R"
```

Reply:
This makes sense as it also orders the assumptions from general to more specific.